workerbee/postgresql.py

```diff
@@ -28,257 +34,328 @@ def check_name(table_name):
+SET time_last_completed=CURRENT_TIMESTAMP, output_data=%(output_data)s, job_duration=%(job_duration)s
+WHERE id=%(id)s
+""".strip()
 
 SET_ROW_CLAIMED_BY_ID_QUERY = r"""
-UPDATE wb__{} SET claimed=true WHERE job_id=%(job_id)s
-"""
+UPDATE {tbl_name} SET time_last_claimed=CURRENT_TIMESTAMP, n_claims = n_claims + 1
+WHERE id=%(id)s
+""".strip()
+
+UPDATE_ROW_N_FAILED_ATTEMPTS_BY_ID_QUERY = r"""
+UPDATE {tbl_name} SET n_failed_attempts = n_failed_attempts + 1
+WHERE id=%(id)s
+""".strip()
 
 TOTAL_ROWS_QUERY = r"""
-SELECT COUNT(*) FROM wb__{0}
-"""
+SELECT COUNT(*) FROM {tbl_name}
+""".strip()
 
 COMPLETED_ROWS_QUERY = r"""
-SELECT COUNT(*) FROM wb__{0} WHERE completed=true
-"""
+SELECT COUNT(*) FROM {tbl_name} WHERE time_last_completed NOTNULL
+""".strip()
 
 
 ###############################################################################
-def table_exists(db_handle, table_id):
-    return db_handle.one(TABLE_EXISTS_QUERY.format(table_id))
 
 
-def setup_table(db_handle, experiment_id, job_ids=None, job_data=None,
-                verbose=False):
-    check_name(experiment_id)
-    does_table_exist = table_exists(db_handle, experiment_id)
+class DBConnectionInfo(object):
 
-    if job_data is None:
-        job_data = {}
-    if job_ids is not None:
-        extra_job_data = set(job_data) - set(job_ids)
-        if len(extra_job_data) > 0:
-            raise ValueError("job_data provided for {} job IDs that aren't in this "
-                             "experiment: {}".format(len(extra_job_data),
-                                                     list(extra_job_data)))
-    if does_table_exist:
-        if verbose:
-            print("Table already exists for experiment '{}'".format(experiment_id))
-            if job_ids is not None:
-                print('Warning: {} provided job IDs were ignored (presumably '
-                      'they are already in the table?)'.format(len(job_ids)))
+    def __init__(self, host=None, port=None, user=None, database=None):
+        self.host = host or os.environ.get('PGHOST', None)
+        self.port = port or os.environ.get('PGPORT', None)
+        self.user = user or os.environ.get('PGUSER', None)
+        self.database = database or os.environ.get('PGDATABASE', None)
+
+    def missing_info(self):
+        return None in {self.host, self.port, self.database, self.user}
+
+    def postgres_connection_string(self):
+        return 'host={} port={} user={} dbname={}'.format(
+            self.host, self.port, self.user, self.database)
+
+    def __eq__(self, other):
+        return (isinstance(other, self.__class__) and
+                self.__dict__ == other.__dict__)
+
+    def __hash__(self):
+        # Ensure that objects with the same connection info will hash the same
+        return hash((self.host, self.port, self.user, self.database))
+
+    def __str__(self):
+        return '{} on {}@{}:{}'.format(self.database, self.user, self.host,
+                                       self.port)
```

Comment:
Unsure how I feel about `DBConnectionInfo`. The change from:
```py
db_info = dict(host='localhost', user='postgres', database='postgres', port=5432)
```
to
```py
from workerbee.postgresql import DBConnectionInfo
db_info = DBConnectionInfo(host='localhost', user='postgres', database='postgres', port=5432)
```
Just adds that little extra friction in my mind. Do like the idea of gathering it into one kwarg through. If we went with the `dict` approach I'd obviously want the verification etc. 

How would you feel about sticking with a dict as the public API but build a `DBConnectionInfo` internally? Could be done at the top of `get_db_handle()`, a try/catch here and we could raise a nice clear error if the user provides a malformed dict?